### PR TITLE
MODELIX-856 Delete its data when a repository is deleted

### PR DESCRIPTION
**Pull request by mhuster23**
Currently we only remove the repository id from the list of available repositories when deleting a repository.
The data still remains in the cache and the database.
With these changes we also delete the objects of a repository from the cache and the database of the model-server, when removing a repository.

## To be verified by reviewers

* [ ] Relevant public API members have been documented
* [ ] Documentation related to this PR is complete
  * [ ] Boundary conditions are documented
  * [ ] Exceptions are documented
  * [ ] Nullability is documented if used
* [ ] Touched existing code has been extended with documentation if missing
* [ ] Code is readable
* [ ] New features and fixed bugs are covered by tests

### CODE REVIEW

File: model-server-test/build.gradle.kts [hunk 1]
```diff
@@ -6,9 +6,12 @@ plugins {
 dependencies {
     testImplementation(kotlin("test"))
     testImplementation(project(":model-server"))
+    testImplementation(project(":model-server").dependencyProject.sourceSets.test.get().runtimeClasspath)
```
Comment (languitar) [inline model-server-test/build.gradle.kts:9]: What does this do? Could benefit from a comment.

File: model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        runBlocking {
+            repoManager.removeRepository(existingRepo)
+            repoManager.removeRepository(repoToBeDeleted)
+        }
+        getDbConnection().use { connection ->
+            connection.prepareStatement("DELETE FROM modelix.model WHERE repository IN (?,?)").apply {
+                setString(1, existingRepo.id)
+                setString(2, repoToBeDeleted.id)
+                execute()
+                close()
```
Comment (languitar) [inline model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt:44]: Why is this `close` needed. Doesn't the prepared statement also implement AutoCloseable? https://docs.oracle.com/javase/tutorial/jdbc/basics/prepared.html

File: model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            }
+        }
+    }
+
+    @Test
+    fun `database does not contain removed data that was not part of the cache`() = runTest {
```
Comment (languitar) [inline model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt:50]: What is "that" here?

File: model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            repoManager.removeRepository(repoToBeDeleted)
+
+            val statement = connection.prepareStatement("SELECT * FROM modelix.model WHERE repository = ?").apply {
+                setString(1, repoToBeDeleted.id)
+            }
+
+            statement.use {
+                val result = it.executeQuery()
+                assertFalse("Database contained leftover repository data.") { result.isBeforeFirst }
+            }
```
Comment (languitar) [inline model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt:70]:
```suggestion
            connection.prepareStatement("SELECT * FROM modelix.model WHERE repository = ?").use {
                setString(1, repoToBeDeleted.id)
                val result = executeQuery()
                assertFalse("Database contained leftover repository data.") { result.isBeforeFirst }
            }
```

Looks a bit weird that the assert block is split up artificially.

File: model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import org.junit.jupiter.api.Test
+import org.modelix.model.lazy.RepositoryId
+import org.modelix.model.server.handlers.RepositoriesManagerTest
+import org.modelix.model.server.store.IgniteStoreClient
+import java.sql.DriverManager
+import kotlin.test.AfterTest
+import kotlin.test.assertFalse
+import kotlin.test.assertTrue
+
+class RepositoriesManagerWithDatabaseTest : RepositoriesManagerTest(IgniteStoreClient()) {
```
Comment (languitar) [inline model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt:27]: This lacks some corner cases:

* Deleting a repo without any data
* Tests for error handling when a query fails

File: model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt [hunk 6]
```diff
@@ -234,12 +236,12 @@ class RepositoriesManager(val client: LocalModelClient) : IRepositoriesManager {
...
                 putVersionHash(repository.getBranchReference(branchName), null)
             }
-            store.getGenericStore().put(branchListKey(repository), null)
+            genericStore.put(branchListKey(repository), null)
             val isolated = checkNotNull(isIsolated(repository)) { "Repository not found: $repository" }
             val existingRepositories = getRepositories(isolated)
             val remainingRepositories = existingRepositories - repository
             store.put(repositoriesListKey(isolated), remainingRepositories.joinToString("\n") { it.id })
-
+            genericStore.removeRepositoryObjects(repository)
```
Comment (languitar) [inline model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt:244]: How expensive is it to do all of this inside the transaction?

File: model-server/src/main/kotlin/org/modelix/model/server/store/IGenericStoreClient.kt [hunk 7]
```diff
@@ -17,8 +17,21 @@
 package org.modelix.model.server.store
 
 import org.modelix.model.IGenericKeyListener
+import org.modelix.model.lazy.RepositoryId
 
-typealias IsolatingStore = IGenericStoreClient<ObjectInRepository>
+interface IsolatingStore : IGenericStoreClient<ObjectInRepository> {
```
[the remark this answers is not in the document]
Comment (languitar) [reply model-server/src/main/kotlin/org/modelix/model/server/store/IGenericStoreClient.kt:22]: Could be addressed in the PR now

File: model-server/src/main/kotlin/org/modelix/model/server/store/IGenericStoreClient.kt [hunk 8]
```diff
@@ -17,8 +17,21 @@
...
+    /**
+     * Default implementation for removing repository objects.
+     * May be overridden by more efficient, store-specific implementations.
```
Comment (languitar) [inline model-server/src/main/kotlin/org/modelix/model/server/store/IGenericStoreClient.kt:25]: Should include valid exceptions to be raised when removing fails.

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt [hunk 9]
```diff
@@ -103,6 +112,43 @@ class IgniteStoreClient(jdbcConfFile: File? = null, inmemory: Boolean = false) :
...
+
+        val filter = IgniteBiPredicate<ObjectInRepository, String?> { key, _ ->
+            key.getRepositoryId() == repositoryId.id
+        }
+        val transformer = IgniteClosure<Cache.Entry<ObjectInRepository, String?>, ObjectInRepository?> { entry ->
+            entry.key
+        }
+        val query = ScanQuery(filter)
+        val toDelete = cache.query(query, transformer).all.asSequence().filterNotNull()
+            .toSortedSet() // sorting is necessary to avoid deadlocks
```
Comment (languitar) [inline model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt:128]: Why can deadlocks occur in case it's not sorted? Some reference to documentation explaining this would be good.

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt [hunk 10]
```diff
@@ -103,6 +112,43 @@ class IgniteStoreClient(jdbcConfFile: File? = null, inmemory: Boolean = false) :
...
+        LOG.info { "Deleting cache entries asynchronously. count=${toDelete.size}" }
+        cache.removeAllAsync(toDelete).listen { LOG.info { "Cache entries deleted." } }
+    }
+
+    private fun removeRepositoryObjectsFromDatabase(repositoryId: RepositoryId) {
+        require(!inmemory) { "Cannot remove from database in inmemory mode." }
```
Comment (languitar) [inline model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt:134]:
```suggestion
        require(!inmemory) { "Cannot remove from database in in-memory mode." }
```

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt [hunk 11]
```diff
@@ -103,6 +112,43 @@ class IgniteStoreClient(jdbcConfFile: File? = null, inmemory: Boolean = false) :
         return cache.associate { it.key to it.value }
     }
 
+    override fun removeRepositoryObjects(repositoryId: RepositoryId) {
+        if (!inmemory) {
+            removeRepositoryObjectsFromDatabase(repositoryId)
+        }
```
Comment (languitar) [inline model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt:118]: It would probably be good to document here why deletion happens twice (when not in in-memory mode) and also why this order of operations is safe (why doesn't ignite try to access the keys that are already deleted now?)

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt [hunk 12]
```diff
@@ -103,6 +112,43 @@ class IgniteStoreClient(jdbcConfFile: File? = null, inmemory: Boolean = false) :
...
+        require(!inmemory) { "Cannot remove from database in inmemory mode." }
+        LOG.info { "Removing repository objects from database." }
+
+        dataSource.connection.use {
+            val statement = it.prepareStatement("DELETE from model WHERE repository = ?").apply {
+                setString(1, repositoryId.id)
+            }
+            try {
+                val deletedRows = statement.executeUpdate()
+                LOG.info { "Deleted rows from database. count=$deletedRows" }
```
Comment (languitar) [inline model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt:143]: The meeting notes for the coding conventions workshop proposed this format: "Transaction failed [transaction=42]"

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt [hunk 13]
```diff
@@ -103,6 +112,43 @@ class IgniteStoreClient(jdbcConfFile: File? = null, inmemory: Boolean = false) :
...
+            } catch (e: SQLException) {
+                LOG.error { e }
+            } finally {
+                statement.close()
```
Comment (languitar) [inline model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt:147]: Should probably be converted to a `use` call instead of explicit `close`.

File: model-server/src/test/kotlin/org/modelix/model/server/handlers/RepositoriesManagerTest.kt [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+import org.modelix.model.server.store.LocalModelClient
+import kotlin.test.AfterTest
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertTrue
+
+class RepositoriesManagerWithInMemoryStoreClientTest : RepositoriesManagerTest(InMemoryStoreClient())
+class RepositoriesManagerWithIgniteStoreClientTest : RepositoriesManagerTest(IgniteStoreClient(inmemory = true))
+
+abstract class RepositoriesManagerTest(val store: IsolatingStore) {
```
[the remark this answers is not in the document]
Comment (languitar) [reply model-server/src/test/kotlin/org/modelix/model/server/handlers/RepositoriesManagerTest.kt:18]: Should we put an explicit suppression explaining why this pattern is chosen (probably to prevent junit from executing the test?)?

File: model-server/src/test/kotlin/org/modelix/model/server/handlers/RepositoriesManagerTest.kt [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    protected val repoManager = RepositoriesManager(LocalModelClient(store))
+    protected fun initRepository(repoId: RepositoryId) {
+        runBlocking {
+            repoManager.createRepository(repoId, "testUser", useRoleIds = true, legacyGlobalStorage = false)
+        }
+    }
```
Comment (languitar) [inline model-server/src/test/kotlin/org/modelix/model/server/handlers/RepositoriesManagerTest.kt:24]: This could simply be a suspend function as it is always called from a coroutine?

File: model-server/src/test/kotlin/org/modelix/model/server/handlers/RepositoriesManagerTest.kt
Comment (languitar) [inline model-server/src/test/kotlin/org/modelix/model/server/handlers/RepositoriesManagerTest.kt:1]: I think the test structure here can be improved, which would also get rid of the abstract test class. The repositories manager is only responsible for invoking the deletion of data. The actual execution of the deletion is the responsibility of each store implementation. Therefore I'd propose:

* Move tests for whether deleting works and prevents deleting other repo's data into tests for each store implementation. Seems we still lack a test for the in-memory store for this...
* Use a mock store in the RepositoriesManagerTest and simply assert that the implementation properly dispatches removal to the store implementation. If we do not have a mocking framework installed yet, I'd propose [mockk](https://mockk.io/).

**Review by languitar — CHANGES_REQUESTED** [review]
Some nitpicks for corner-cases and documentation and a proposal for an improved test structure. Apart from that, this generally looks fine to me.

File: model-server/src/main/kotlin/org/modelix/model/server/store/IGenericStoreClient.kt (hunk 8 above)
Comment (mhuster23) [reply model-server/src/main/kotlin/org/modelix/model/server/store/IGenericStoreClient.kt:25]: We currently don't explicitly throw custom exceptions in any of our StoreClients. We should probably add some to wrap store-specific ones (ignite and sql exceptions) but that would require [MODELIX-799](https://issues.modelix.org/issue/MODELIX-799/Create-a-proposal-on-how-to-organize-exceptions). After that we should apply it to this and the already existing operations.

File: model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt (hunk 6 above)
Comment (mhuster23) [reply model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt:244]: I don't know how to quantify that. In the long run repository objects should be stored in a separate cache without transactions, as the data is conceptually immutable (but removable).

Comment (languitar) [reply model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt:244]: How long does it take for our big customer? The question is if it's valid to do all that in the transaction and also in the scope of the REST request.

File: model-server/src/test/kotlin/org/modelix/model/server/handlers/RepositoriesManagerTest.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (mhuster23) [reply model-server/src/test/kotlin/org/modelix/model/server/handlers/RepositoriesManagerTest.kt]: outdated

File: model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt (hunk 2 above)
Comment (mhuster23) [reply model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt]: fixed

File: model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt (hunk 3 above)
Comment (mhuster23) [reply model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt]: outdated

File: model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt (hunk 4 above)
Comment (mhuster23) [reply model-server-test/src/test/kotlin/RepositoriesManagerWithDatabaseTest.kt]: fixed

File: model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt (hunk 6 above)
Comment (mhuster23) [reply model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt:244]: Haven't run into issues yet. We can follow up on it, if it becomes a problem.

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt (hunk 9 above)
Comment (mhuster23) [reply model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt]: added a hint

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt (hunk 10 above)
Comment (mhuster23) [reply model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt]: fixed

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt (hunk 11 above)
Comment (mhuster23) [reply model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt]: done

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt (hunk 12 above)
Comment (mhuster23) [reply model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt]: done

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt (hunk 13 above)
Comment (mhuster23) [reply model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt]: done

File: model-server/src/test/kotlin/org/modelix/model/server/handlers/RepositoriesManagerTest.kt (hunk 15 above)
Comment (mhuster23) [reply model-server/src/test/kotlin/org/modelix/model/server/handlers/RepositoriesManagerTest.kt]: fixed

File: model-server/src/test/kotlin/org/modelix/model/server/handlers/RepositoriesManagerTest.kt
Comment (mhuster23) [reply model-server/src/test/kotlin/org/modelix/model/server/handlers/RepositoriesManagerTest.kt:1]: done

File: model-server-test/src/test/kotlin/IgnitePostgresRepositoryRemovalTest.kt [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+    @Test
+    fun `removal does not affect other repository data in the cache`() {
+        val existingEntries = mapOf(
+            ObjectInRepository(existing.id, "key0") to "value0",
+            ObjectInRepository(existing.id, "key1") to "value1",
+        )
+        store.putAll(existingEntries)
+
+        val toDeleteId = RepositoryId("cache-other-removal")
```
[the remark this answers is not in the document]
Comment (languitar) [reply model-server-test/src/test/kotlin/IgnitePostgresRepositoryRemovalTest.kt:110]: This variable can probably be removed then

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt [hunk 17]
```diff
@@ -18,19 +18,25 @@
...
 import java.io.IOException
+import java.sql.SQLException
 import java.util.*
+import javax.cache.Cache
 import javax.sql.DataSource
 
 private val LOG = KotlinLogging.logger { }
 
-class IgniteStoreClient(jdbcConfFile: File? = null, inmemory: Boolean = false) : IsolatingStore, AutoCloseable {
+class IgniteStoreClient(jdbcConfFile: File? = null, private val inmemory: Boolean = false) : IsolatingStore, AutoCloseable {
```
[the remark this answers is not in the document]
Comment (languitar) [reply model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt:39]: > Touched existing code has been extended with documentation if missing

Can we address this here?

File: model-server/src/main/kotlin/org/modelix/model/server/store/InMemoryStoreClient.kt [hunk 18]
```diff
@@ -103,7 +103,12 @@
                 try {
                     transactionValues = HashMap()
                     val result = body()
-                    values.putAll(transactionValues!!)
+                    val puts = transactionValues!!.filterValues { it != null }
```
[the remark this answers is not in the document]
Comment (languitar) [reply model-server/src/main/kotlin/org/modelix/model/server/store/InMemoryStoreClient.kt:106]: How can `transactionValues` actually ever be `null` here?

File: model-server-test/src/test/kotlin/IgnitePostgresRepositoryRemovalTest.kt [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+import org.modelix.model.server.store.IgniteStoreClient
+import org.modelix.model.server.store.ObjectInRepository
+import java.sql.DriverManager
+import kotlin.test.AfterTest
+import kotlin.test.assertEquals
+import kotlin.test.assertFalse
+import kotlin.test.assertTrue
+
+class IgnitePostgresRepositoryRemovalTest {
+    private fun getDbConnection() = DriverManager.getConnection(System.getProperty("jdbc.url"), "modelix", "modelix")
```
Comment (languitar) [inline model-server-test/src/test/kotlin/IgnitePostgresRepositoryRemovalTest.kt:28]: Why do we create a new connection on every access? Could be kept alive for all test cases.

File: model-server/src/main/kotlin/org/modelix/model/server/store/IGenericStoreClient.kt (hunk 8 above)
Comment (languitar) [reply model-server/src/main/kotlin/org/modelix/model/server/store/IGenericStoreClient.kt:29]: Not totally happy to ignore this right now as the other task will grow with every change we make this way. But ok

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt [hunk 20]
```diff
@@ -103,6 +112,44 @@ class IgniteStoreClient(jdbcConfFile: File? = null, inmemory: Boolean = false) :
         return cache.associate { it.key to it.value }
     }
 
+    override fun removeRepositoryObjects(repositoryId: RepositoryId) {
+        if (!inmemory) {
+            // Not all entries are in the cache. We delete them directly instead of loading them into the cache first.
+            // This should be safe as the repository has already been removed from the list of available ones.
```
Comment (languitar) [inline model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt:118]: Let's explicitly document this contract in the interface kdoc for `removeRepositoryObjects` a la: "callers need to ensure that repositories are not usable anymore before calling this method". This increases the chance that this fact is not forgotten during the next refactoring.

**Review by languitar — DISMISSED** [review]
A few nit-picks and doc improvements. But otherwise that looks fine to me and even more concise now.

File: model-server/src/main/kotlin/org/modelix/model/server/store/InMemoryStoreClient.kt (hunk 18 above)
[the remark this answers is not in the document]
Comment (mhuster23) [reply model-server/src/main/kotlin/org/modelix/model/server/store/InMemoryStoreClient.kt:106]: the `body` lambda could set it to `null` as it is a global variable

File: model-server-test/src/test/kotlin/IgnitePostgresRepositoryRemovalTest.kt (hunk 19 above)
Comment (mhuster23) [reply model-server-test/src/test/kotlin/IgnitePostgresRepositoryRemovalTest.kt]: fixed

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt (hunk 20 above)
Comment (mhuster23) [reply model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt:122]: fixed

File: model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt (hunk 17 above)
[the remark this answers is not in the document]
Comment (mhuster23) [reply model-server/src/main/kotlin/org/modelix/model/server/store/IgniteStoreClient.kt:43]: done

File: model-server/src/main/kotlin/org/modelix/model/server/store/InMemoryStoreClient.kt (hunk 18 above)
[the remark this answers is not in the document]
Comment (languitar) [reply model-server/src/main/kotlin/org/modelix/model/server/store/InMemoryStoreClient.kt:106]: so we should probably use `requireNotNull` then in case this must not happen.

**Comment by languitar** [comment]
I can't find the one conversation that's supposed to still be open. The GitHub UI doesn't show it in the discussions and cannot navigate to it in the code review.

**Review by languitar — DISMISSED** [review]

**Review by languitar — APPROVED** [review]
